File: lib/product-units.ts
import type { InventoryItem, Product } from "./types";

export function getStockUnit(product?: Product) {
    return product?.stockUnit?.trim() || product?.unit?.trim() || "واحد";
}

export function getOrderUnit(product?: Product) {
    return product?.orderUnit?.trim() || getStockUnit(product);
}

export function getOrderUnitQuantity(product?: Product) {
    const value = product?.orderUnitQuantity ?? 1;
    return Number.isFinite(value) && value > 0 ? value : 1;
}

export function getOrderQuantityStep(product?: Product) {
    const value = product?.orderQuantityStep ?? 0.001;
    return Number.isFinite(value) && value > 0 ? value : 0.001;
}

export function orderToStockQuantity(product: Product | undefined, orderQuantity: number) {
    const safeQuantity = Number.isFinite(orderQuantity) ? Math.max(0, orderQuantity) : 0;
    return Math.round(safeQuantity * getOrderUnitQuantity(product) * 1000) / 1000;
}

export function stockToOrderQuantity(product: Product | undefined, stockQuantity: number) {
    const safeQuantity = Number.isFinite(stockQuantity) ? Math.max(0, stockQuantity) : 0;
    const rawOrderQuantity = safeQuantity / getOrderUnitQuantity(product);
    const step = getOrderQuantityStep(product);

    if (step >= 1) {
        return Math.floor(rawOrderQuantity / step) * step;
    }

    return Math.floor(rawOrderQuantity / step) * step;
}

export function normalizeOrderQuantity(product: Product | undefined, quantity: number) {
    const safeQuantity = Number.isFinite(quantity) ? Math.max(0, quantity) : 0;
    const step = getOrderQuantityStep(product);
    const normalized = step >= 1
        ? Math.floor(safeQuantity / step) * step
        : Math.round(safeQuantity / step) * step;

    return Math.round(normalized * 1000) / 1000;
}

export function formatNumber(value: number) {
    return new Intl.NumberFormat("en-US", { maximumFractionDigits: 3 }).format(value);
}

export function formatStockQuantity(product: Product | undefined, quantity: number) {
    return `${formatNumber(quantity)} ${getStockUnit(product)}`;
}

export function formatOrderQuantity(product: Product | undefined, quantity: number) {
    return `${formatNumber(quantity)} ${getOrderUnit(product)}`;
}

export function formatUnitConversion(product: Product | undefined) {
    const orderQuantity = getOrderUnitQuantity(product);
    const orderUnit = getOrderUnit(product);
    const stockUnit = getStockUnit(product);

    if (orderUnit === stockUnit && orderQuantity === 1) return `${stockUnit}`;

    return `1 ${orderUnit} = ${formatNumber(orderQuantity)} ${stockUnit}`;
}

export function hasUnitConversion(product?: Product) {
    return getOrderUnit(product) !== getStockUnit(product) || getOrderUnitQuantity(product) !== 1;
}

export function getOrderRemainder(product: Product | undefined, stockQuantity: number) {
    const safeQuantity = Number.isFinite(stockQuantity) ? Math.max(0, stockQuantity) : 0;
    const wholeOrderUnits = stockToOrderQuantity(product, safeQuantity);
    const consumedStock = orderToStockQuantity(product, wholeOrderUnits);

    return Math.max(0, Math.round((safeQuantity - consumedStock) * 1000) / 1000);
}

/**
 * "قابل درخواست" as whole order-units alone hides any leftover stock that
 * doesn't add up to one more package (e.g. 2.35kg of a 3.5kg-per-بسته item
 * reads as "0 بسته" with no sign the 2.35kg exists). Append the remainder in
 * stock units whenever it's non-zero so partial stock stays visible.
 */
export function formatAvailableQuantity(product: Product | undefined, stockQuantity: number) {
    const orderQuantity = stockToOrderQuantity(product, stockQuantity);
    const base = formatOrderQuantity(product, orderQuantity);

    if (!hasUnitConversion(product)) return base;

    const remainder = getOrderRemainder(product, stockQuantity);
    if (remainder <= 0) return base;

    return `${base} (${formatStockQuantity(product, remainder)} باقیمانده)`;
}

export type StockStatusTier = "out" | "critical" | "low" | "ok";

export function getStockStatus(product: Product | undefined, inventoryItem?: InventoryItem): StockStatusTier {
    const quantity = inventoryItem?.currentQuantity ?? 0;
    const availableOrderQuantity = stockToOrderQuantity(product, quantity);

    if (availableOrderQuantity <= 0) return "out";
    if (quantity <= (inventoryItem?.criticalQuantity ?? 0)) return "critical";
    if (quantity <= (inventoryItem?.minimumQuantity ?? 0)) return "low";

    return "ok";
}

export function getStockStatusLabel(tier: StockStatusTier) {
    if (tier === "out") return "ناموجود";
    if (tier === "critical") return "موجودی بحرانی";
    if (tier === "low") return "موجودی کم";

    return "موجود";
}

export function getStockStatusStyle(tier: StockStatusTier) {
    if (tier === "out") return "bg-red-50 text-red-700 ring-red-100";
    if (tier === "critical") return "bg-orange-50 text-orange-700 ring-orange-100";
    if (tier === "low") return "bg-yellow-50 text-yellow-700 ring-yellow-100";

    return "bg-[#e0ffe0] text-[#007A00] ring-[#9ee89e]";
}

File: android/app/src/main/java/cafe/managment/data/util/ProductUnits.kt
package cafe.managment.data.util

import cafe.managment.data.model.InventoryItem
import cafe.managment.data.model.Product
import java.text.NumberFormat
import java.util.Locale
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.round

/** Kotlin port of staff-management-saas/lib/product-units.ts — keep in sync with that file. */
object ProductUnits {

    private val numberFormat: NumberFormat = NumberFormat.getNumberInstance(Locale.US).apply {
        maximumFractionDigits = 3
        isGroupingUsed = true
    }

    fun stockUnit(product: Product?): String =
        product?.stockUnit?.trim()?.takeIf { it.isNotEmpty() }
            ?: product?.unit?.trim()?.takeIf { it.isNotEmpty() }
            ?: "واحد"

    fun orderUnit(product: Product?): String =
        product?.orderUnit?.trim()?.takeIf { it.isNotEmpty() } ?: stockUnit(product)

    fun orderUnitQuantity(product: Product?): Double {
        val value = product?.orderUnitQuantity ?: 1.0
        return if (value.isFinite() && value > 0) value else 1.0
    }

    fun orderQuantityStep(product: Product?): Double {
        val value = product?.orderQuantityStep ?: 0.001
        return if (value.isFinite() && value > 0) value else 0.001
    }

    fun orderToStockQuantity(product: Product?, orderQuantity: Double): Double {
        val safeQuantity = if (orderQuantity.isFinite()) max(0.0, orderQuantity) else 0.0
        return round(safeQuantity * orderUnitQuantity(product) * 1000) / 1000
    }

    fun stockToOrderQuantity(product: Product?, stockQuantity: Double): Double {
        val safeQuantity = if (stockQuantity.isFinite()) max(0.0, stockQuantity) else 0.0
        val rawOrderQuantity = safeQuantity / orderUnitQuantity(product)
        val step = orderQuantityStep(product)
        return floor(rawOrderQuantity / step) * step
    }

    fun normalizeOrderQuantity(product: Product?, quantity: Double): Double {
        val safeQuantity = if (quantity.isFinite()) max(0.0, quantity) else 0.0
        val step = orderQuantityStep(product)
        val normalized = if (step >= 1) {
            floor(safeQuantity / step) * step
        } else {
            round(safeQuantity / step) * step
        }
        return round(normalized * 1000) / 1000
    }

    fun formatNumber(value: Double): String = numberFormat.format(value)

    fun formatStockQuantity(product: Product?, quantity: Double): String =
        "${formatNumber(quantity)} ${stockUnit(product)}"

    fun formatOrderQuantity(product: Product?, quantity: Double): String =
        "${formatNumber(quantity)} ${orderUnit(product)}"

    fun formatUnitConversion(product: Product?): String {
        val quantity = orderUnitQuantity(product)
        val order = orderUnit(product)
        val stock = stockUnit(product)
        if (order == stock && quantity == 1.0) return stock
        return "1 $order = ${formatNumber(quantity)} $stock"
    }

    fun hasUnitConversion(product: Product?): Boolean =
        orderUnit(product) != stockUnit(product) || orderUnitQuantity(product) != 1.0

    fun getOrderRemainder(product: Product?, stockQuantity: Double): Double {
        val safeQuantity = if (stockQuantity.isFinite()) max(0.0, stockQuantity) else 0.0
        val wholeOrderUnits = stockToOrderQuantity(product, safeQuantity)
        val consumedStock = orderToStockQuantity(product, wholeOrderUnits)
        return max(0.0, round((safeQuantity - consumedStock) * 1000) / 1000)
    }

    fun formatAvailableQuantity(product: Product?, stockQuantity: Double): String {
        val orderQuantity = stockToOrderQuantity(product, stockQuantity)
        val base = formatOrderQuantity(product, orderQuantity)
        if (!hasUnitConversion(product)) return base

        val remainder = getOrderRemainder(product, stockQuantity)
        if (remainder <= 0) return base

        return "$base (${formatStockQuantity(product, remainder)} باقیمانده)"
    }

    enum class StockStatusTier { OUT, CRITICAL, LOW, OK }

    fun getStockStatus(product: Product?, inventoryItem: InventoryItem?): StockStatusTier {
        val quantity = inventoryItem?.currentQuantity ?: 0.0
        val availableOrderQuantity = stockToOrderQuantity(product, quantity)

        return when {
            availableOrderQuantity <= 0 -> StockStatusTier.OUT
            quantity <= (inventoryItem?.criticalQuantity ?: 0.0) -> StockStatusTier.CRITICAL
            quantity <= (inventoryItem?.minimumQuantity ?: 0.0) -> StockStatusTier.LOW
            else -> StockStatusTier.OK
        }
    }

    fun getStockStatusLabel(tier: StockStatusTier): String = when (tier) {
        StockStatusTier.OUT -> "ناموجود"
        StockStatusTier.CRITICAL -> "موجودی بحرانی"
        StockStatusTier.LOW -> "موجودی کم"
        StockStatusTier.OK -> "موجود"
    }
}
